File: apps/posting/posting_core/controller/albums.py
from __future__ import annotations

import json
import time
from datetime import datetime

from posting_core.controller.config import ALBUM_SETTLE_SECONDS, api, log
from posting_core.controller.db import db, set_state, update_draft
from posting_core.controller.drafts import create_draft
from posting_core.controller.media import media_json_value
from posting_core.controller.ui import send_preview
from posting_core.time_utils import now_iso


def album_key(admin_id, chat_id, media_group_id, action=None, draft_id=None):
    return f"{admin_id}:{chat_id}:{media_group_id}:{action or 'draft'}:{draft_id or ''}"


def append_pending_album(admin_id, chat_id, media_group_id, text_ru, media, action=None, draft_id=None, entities=None):
    if not media:
        return
    key = album_key(admin_id, chat_id, media_group_id, action, draft_id)
    with db() as conn:
        row = conn.execute("SELECT * FROM pending_albums WHERE id=?", (key,)).fetchone()
        if row:
            items = json.loads(row["media_json"])
            items.append(media)
            new_text = text_ru or row["text_ru"] or ""
            entities_json = json.dumps(entities or json.loads(row["text_entities_json"] or "[]"), ensure_ascii=False)
            conn.execute(
                "UPDATE pending_albums SET text_ru=?, text_entities_json=?, media_json=?, updated_at=? WHERE id=?",
                (new_text, entities_json, json.dumps(items, ensure_ascii=False), now_iso(), key),
            )
            notified = row["notified"]
        else:
            conn.execute(
                "INSERT INTO pending_albums(id,admin_id,chat_id,media_group_id,action,draft_id,text_ru,text_entities_json,media_json,notified,updated_at) VALUES(?,?,?,?,?,?,?,?,?,?,?)",
                (
                    key,
                    admin_id,
                    chat_id,
                    str(media_group_id),
                    action,
                    draft_id,
                    text_ru or "",
                    json.dumps(entities or [], ensure_ascii=False),
                    json.dumps([media], ensure_ascii=False),
                    0,
                    now_iso(),
                ),
            )
            notified = 0
        conn.commit()
    if not notified:
        api(
            "sendMessage",
            {"chat_id": chat_id, "text": "Album received. I will create/update the draft in a few seconds."},
        )
        with db() as conn:
            conn.execute("UPDATE pending_albums SET notified=1 WHERE id=?", (key,))
            conn.commit()


def finalize_pending_albums():
    deadline = time.time() - ALBUM_SETTLE_SECONDS
    with db() as conn:
        rows = [dict(row) for row in conn.execute("SELECT * FROM pending_albums")]
    for row in rows:
        try:
            updated_ts = datetime.fromisoformat(row["updated_at"]).timestamp()
        except Exception:
            updated_ts = 0
        if updated_ts > deadline:
            continue
        media_items = json.loads(row["media_json"])
        try:
            if row.get("action") in {"replace_ru_media", "replace_en_media"} and row.get("draft_id"):
                media_field = "media_ru_json" if row.get("action") == "replace_ru_media" else "media_en_json"
                update_draft(
                    row["draft_id"], **{media_field: json.dumps(media_json_value(media_items), ensure_ascii=False)}
                )
                from posting_core.controller.db import get_draft
                from posting_core.controller.schedule import rebalance_all_scheduled_drafts

                if (get_draft(row["draft_id"]) or {}).get("status") == "scheduled":
                    rebalance_all_scheduled_drafts()
                set_state(row["admin_id"])
                send_preview(row["chat_id"], row["draft_id"])
            else:
                draft_id = create_draft(
                    row["admin_id"],
                    row.get("text_ru") or "",
                    media_items,
                    json.loads(row.get("text_entities_json") or "[]"),
                )
                send_preview(row["chat_id"], draft_id)
            with db() as conn:
                conn.execute("DELETE FROM pending_albums WHERE id=?", (row["id"],))
                conn.commit()
        except Exception as exc:
            log(f"failed to finalize album {row['id']}: {exc}")

File: apps/backend/src/bot/albums.ts
import { and, asc, eq, lte } from "drizzle-orm";
import type { Bot } from "grammy";
import { parseArrayValue } from "../content/message.js";
import { type BackendDb, unsafeDb } from "../db/client.js";
import { pendingAlbums } from "../db/schema.js";
import type { BackendConfig } from "../foundation/config.js";
import { t } from "../foundation/i18n/index.js";
import { log } from "../foundation/logger.js";
import { setTelegramPostCard } from "../interfaces/telegram/control-cards.js";
import { importTelegramAlbumMedia } from "../interfaces/telegram/media-ingress.js";
import { createStudioServices } from "../studio/services/index.js";
import { botLocale } from "./i18n.js";
import { clearPostAdminStateIfCurrent, getPostAdminState } from "./post-state.js";
import { translatePostText } from "./post-translation.js";
import { draftPreview } from "./preview.js";

// pending_albums.notified lifecycle: an album is SETTLED once its caption and
// media are collected, then CLAIMED by exactly one worker before finalization.
const ALBUM_SETTLED = 1;
const ALBUM_CLAIMED = 2;
// A claim is durable so a process crash cannot lose the album, but it needs a
// lease so a claim abandoned mid-import can be picked up by the next cycle.
const ALBUM_CLAIM_LEASE_MS = 10 * 60_000;
// A failed finalization goes back to SETTLED and is retried once per settle
// window. Deterministic failures (an expired file_id, a draft deleted mid-flight)
// would loop forever, so give up after a few tries and tell the sender instead
// of retrying silently at ~1 Hz.
const ALBUM_MAX_ATTEMPTS = 5;

type PendingAlbumInput = {
  actorId: number;
  chatId: number;
  mediaGroupId: string;
  text: string;
  entities: unknown[];
  media: Record<string, unknown>;
  action: string | null;
  draftId: number | null;
  stateRevision: number | null;
};

export function appendPendingAlbum(backendDb: BackendDb, input: PendingAlbumInput): boolean {
  const id = `${input.actorId}:${input.chatId}:${input.mediaGroupId}:${input.action ?? "draft"}:${input.draftId ?? ""}`;
  const row = unsafeDb(backendDb)
    .db.select({ mediaJson: pendingAlbums.mediaJson, textRu: pendingAlbums.textRu, textEntitiesJson: pendingAlbums.textEntitiesJson })
    .from(pendingAlbums)
    .where(eq(pendingAlbums.id, id))
    .get();
  const media = row ? parseArrayValue(row.mediaJson) : [];
  media.push(input.media);
  const now = new Date().toISOString();
  const values = {
    id,
    actorId: input.actorId,
    chatId: input.chatId,
    mediaGroupId: input.mediaGroupId,
    action: input.action,
    draftId: input.draftId,
    stateRevision: input.stateRevision,
    textRu: input.text || row?.textRu || "",
    textEntitiesJson: JSON.stringify(input.entities.length ? input.entities : parseArrayValue(row?.textEntitiesJson)),
    mediaJson: JSON.stringify(media),
    notified: ALBUM_SETTLED,
    updatedAt: now,
  };
  unsafeDb(backendDb)
    .db.insert(pendingAlbums)
    .values(values)
    .onConflictDoUpdate({
      target: pendingAlbums.id,
      set: {
        textRu: values.textRu,
        textEntitiesJson: values.textEntitiesJson,
        mediaJson: values.mediaJson,
        notified: ALBUM_SETTLED,
        updatedAt: now,
      },
    })
    .run();
  return !row;
}

export async function finalizePendingAlbums(bot: Bot | null, backendDb: BackendDb, config: BackendConfig): Promise<number> {
  if (!bot) return 0;
  const now = new Date();
  const nowIso = now.toISOString();
  const claimCutoff = new Date(now.getTime() - ALBUM_CLAIM_LEASE_MS).toISOString();
  // Recover claims left behind by a crashed worker. Keep updatedAt untouched:
  // the row can be selected in this same cycle if its settle window elapsed.
  unsafeDb(backendDb)
    .db.update(pendingAlbums)
    .set({ notified: ALBUM_SETTLED })
    .where(and(eq(pendingAlbums.notified, ALBUM_CLAIMED), lte(pendingAlbums.updatedAt, claimCutoff)))
    .run();
  const cutoff = new Date(now.getTime() - config.CONTROLLER_ALBUM_SETTLE_SECONDS * 1000).toISOString();
  const rows = unsafeDb(backendDb)
    .db.select({
      id: pendingAlbums.id,
      actorId: pendingAlbums.actorId,
      chatId: pendingAlbums.chatId,
      action: pendingAlbums.action,
      draftId: pendingAlbums.draftId,
      stateRevision: pendingAlbums.stateRevision,
      attemptCount: pendingAlbums.attemptCount,
      textRu: pendingAlbums.textRu,
      textEntitiesJson: pendingAlbums.textEntitiesJson,
      mediaJson: pendingAlbums.mediaJson,
    })
    .from(pendingAlbums)
    .where(and(eq(pendingAlbums.notified, ALBUM_SETTLED), lte(pendingAlbums.updatedAt, cutoff)))
    .orderBy(asc(pendingAlbums.updatedAt))
    .all();
  let completed = 0;
  for (const row of rows) {
    const claim = unsafeDb(backendDb)
      .db.update(pendingAlbums)
      .set({ notified: ALBUM_CLAIMED, updatedAt: nowIso })
      .where(and(eq(pendingAlbums.id, row.id), eq(pendingAlbums.notified, ALBUM_SETTLED), lte(pendingAlbums.updatedAt, cutoff)))
      .returning({ id: pendingAlbums.id })
      .get();
    if (!claim) continue;
    try {
      if (row.stateRevision != null && getPostAdminState(backendDb, row.actorId)?.revision !== row.stateRevision) {
        unsafeDb(backendDb)
          .db.delete(pendingAlbums)
          .where(and(eq(pendingAlbums.id, row.id), eq(pendingAlbums.notified, ALBUM_CLAIMED)))
          .run();
        log("warn", "stale album discarded", { album: row.id, actorId: row.actorId, stateRevision: row.stateRevision });
        continue;
      }
      const media = await importTelegramAlbumMedia(bot, backendDb, config, row.actorId, parseArrayValue(row.mediaJson));
      const draftId = row.draftId;
      const isEdit = row.action === "edit_ru" || row.action === "edit_en";
      const isMediaReplacement = row.action === "replace_ru_media" || row.action === "replace_en_media";
      if ((isEdit || isMediaReplacement) && draftId) {
        createStudioServices(backendDb, config).posts.edit(row.actorId, draftId, {
          locale: row.action === "edit_ru" || row.action === "replace_ru_media" ? "ru" : "en",
          text: isMediaReplacement ? "" : row.textRu,
          entities: isMediaReplacement ? [] : parseArrayValue(row.textEntitiesJson),
          media,
          ...(isMediaReplacement ? { replaceMediaOnly: true } : {}),
        });
        clearPostAdminStateIfCurrent(backendDb, row.actorId, row.action, draftId, row.stateRevision);
        await refreshDraftControlCard(bot, backendDb, config, row.actorId, draftId, row.chatId);
      } else {
        const text = row.textRu;
        const textEn = await translatePostText(text, config);
        const created = createStudioServices(backendDb, config).publications.create(row.actorId, {
          kind: "post",
          message: { text, textEn, media, entities: parseArrayValue(row.textEntitiesJson) },
        }).id;
        await refreshDraftControlCard(bot, backendDb, config, row.actorId, created, row.chatId);
        clearPostAdminStateIfCurrent(backendDb, row.actorId, row.action, row.draftId, row.stateRevision);
      }
      const removed = unsafeDb(backendDb)
        .db.delete(pendingAlbums)
        .where(and(eq(pendingAlbums.id, row.id), eq(pendingAlbums.notified, ALBUM_CLAIMED)))
        .returning({ id: pendingAlbums.id })
        .get();
      if (removed) completed += 1;
    } catch (error) {
      const attempts = row.attemptCount + 1;
      const exhausted = attempts >= ALBUM_MAX_ATTEMPTS;
      if (exhausted) {
        unsafeDb(backendDb).db.delete(pendingAlbums).where(eq(pendingAlbums.id, row.id)).run();
        await notifyAlbumGaveUp(bot, backendDb, row.actorId, row.chatId);
      } else {
        unsafeDb(backendDb)
          .db.update(pendingAlbums)
          .set({ notified: ALBUM_SETTLED, attemptCount: attempts, updatedAt: new Date().toISOString() })
          .where(and(eq(pendingAlbums.id, row.id), eq(pendingAlbums.notified, ALBUM_CLAIMED)))
          .run();
      }
      log(exhausted ? "error" : "warn", "album finalization failed", {
        album: row.id,
        attempts,
        exhausted,
        error: String(error),
      });
    }
  }
  return completed;
}

/** Last word on an album that will never become a draft. Best-effort: a failed
 * notification must not resurrect the row we just dropped. */
async function notifyAlbumGaveUp(bot: Bot, backendDb: BackendDb, actorId: number, chatId: number): Promise<void> {
  try {
    await bot.api.sendMessage(chatId, t(botLocale(backendDb, actorId), "post.album-failed"));
  } catch (error) {
    log("warn", "album give-up notice failed", { chat: chatId, error: String(error) });
  }
}

async function refreshDraftControlCard(
  bot: Bot,
  backendDb: BackendDb,
  config: BackendConfig,
  _actorId: number,
  draftId: number,
  chatId: number,
): Promise<void> {
  const preview = draftPreview(backendDb, draftId, config);
  // A completed chat edit gets a fresh card at the bottom. Previous cards are
  // history, never a moving conversation prompt above the user's reply.
  const control = await bot.api.sendMessage(chatId, preview.text, { parse_mode: "Markdown", reply_markup: preview.keyboard });
  setTelegramPostCard(backendDb, draftId, chatId, control.message_id);
}
